pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -51,21 +60,70 @@ type ReaderPool struct {
 	lazyReaders   map[*LazyBinaryReader]struct{}
 }
 
-// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers if enabled.
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+// LazyLoadedHeadersSnapshotConfig stores information needed to track lazy loaded index headers.
+type LazyLoadedHeadersSnapshotConfig struct {
+	Path   string
+	UserID string
+}
+
+type lazyLoadedHeadersSnapshot struct {
+	// HeaderLastUsedTime is map of index header ulid.ULID to timestamp in millisecond.
+	HeaderLastUsedTime map[ulid.ULID]int64 `json:"header_last_used_time"`
+	UserID             string              `json:"user_id"`
+}
+
+// persist atomically writes this snapshot to persistDir.
+func (l lazyLoadedHeadersSnapshot) persist(persistDir string) error {
+	// Create temporary path for fsync.
+	// We don't use temporary folder because the process might not have access to the temporary folder.
+	tmpPath := filepath.Join(persistDir, strings.Join([]string{"tmp", lazyLoadedHeadersListFile}, "-"))
+
+	data, err := json.Marshal(l)
+	if err != nil {
+		return err
+	}
+
+	if err := atomicfs.CreateFile(tmpPath, bytes.NewReader(data)); err != nil {
+		return err
+	}
+	defer os.Remove(tmpPath)
+
+	// move the written file to the actual path
+	finalPath := filepath.Join(persistDir, lazyLoadedHeadersListFile)
+	// we rely on the atomicity of this on Unix systems for this method to behave correctly
+	return os.Rename(tmpPath, finalPath)
```

Comment:
The logic of writing a tmp file first and the moving it looks like something that should be moved to the new `atomicfs` package (e.g. `atomicfs.CreateFileAndMove()`).